### PR TITLE
Contract maven plugin - incremental test generation

### PR DESCRIPTION
**Pull request by michaldo**

### CODE REVIEW

File: spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/pom.xml [hunk 1]
```diff
@@ -367,6 +367,11 @@
 			<artifactId>plexus-archiver</artifactId>
 			<version>${plexus-archiver.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>org.apache.maven.shared</groupId>
+			<artifactId>maven-shared-incremental</artifactId>
+			<version>1.1</version>
```
[the remark this answers is not in the document]
Comment (michaldo) [reply spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/pom.xml:373]: sure, I will do

File: spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/src/main/java/org/springframework/cloud/contract/maven/verifier/GenerateTestsMojo.java [hunk 2]
```diff
@@ -367,4 +411,33 @@ public void setAssertJsonSize(boolean assertJsonSize) {
...
+	/**
+	 * Prevents against scenario: 1. Contract is added 2. Test is generated 3. Contract is
+	 * deleted 4. Generated test still exists
+	 */
+	private final class LeftOverPrevention {
+
+		private IncrementalBuildHelper incrementalBuildHelper = new IncrementalBuildHelper(
+				mojoExecution, session);
+
+		private File generatedDirectory;
```
[the remark this answers is not in the document]
Comment (michaldo) [reply spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/src/main/java/org/springframework/cloud/contract/maven/verifier/GenerateTestsMojo.java:423]: sure, I will do

**Comment by michaldo** [comment]
About plus:

> Shouldn't we do the same for stubs?

Maybe. I made change for Java code generation, because lack of incrementalism trigger all test recompilation each build - and compilation is costly.

Going back to stubs: do you mean both goals `convert` and `generateStubs` should be incremental? Regardless their cost, it would be elegant to have incrementalism implemented for both. I already implemented that, so if you confirm both goals should be modified, I will update PR.

_BTW, do you prefer single commit PR or multiple commits (and you will stash them on your side)_

> Can you please add some tests?

Well, PR is non-functional change, so existing tests should be enough

**Comment by marcingrzejszczak** [comment]
> Going back to stubs: do you mean both goals convert and generateStubs should be incremental? Regardless their cost, it would be elegant to have incrementalism implemented for both. I already implemented that, so if you confirm both goals should be modified, I will update PR.

Correct, thanks a lot.

> BTW, do you prefer single commit PR or multiple commits (and you will stash them on your side)

Do however you please, I'll stash them upon merge.

> Well, PR is non-functional change, so existing tests should be enough

:+1:

**Comment by marcingrzejszczak** [comment]
Great work @michaldo , thanks a lot!

**Pull request merged**